container-engine-lib/lib/backend_impls/docker/object_attributes_provider/enclave_object_attributes_provider.go

```diff
@@ -516,6 +528,94 @@ func (provider *dockerEnclaveObjectAttributesProviderImpl) getLabelsForEnclaveOb
 	return labels, nil
 }
 
+// Return Traefik labels
+// Including the labels required to route traffic to the user service ports based on the Host header:
+// <port number>-<service short uuid>-<enclave short uuid>
+// The Traefik service name format is: <enclave short uuid>-<service short uuid>-<port number>
+// With the following input:
+//   Enclave short UUID: 65d2fb6d6732
+//   Service short UUID: 3771c85af16a
+//   HTTP Port 1 number: 80
+//   HTTP Port 2 number: 81
+// the following labels are returned:
+//   "traefik.enable": "true",
+//   "traefik.http.routers.65d2fb6d6732-3771c85af16a-80.rule": "Host(`80-3771c85af16a-65d2fb6d6732`)",
+//   "traefik.http.routers.65d2fb6d6732-3771c85af16a-80.service": "65d2fb6d6732-3771c85af16a-80",
+//   "traefik.http.services.65d2fb6d6732-3771c85af16a-80.loadbalancer.server.port": "80"
+//   "traefik.http.routers.65d2fb6d6732-3771c85af16a-81.rule": "Host(`81-3771c85af16a-65d2fb6d6732`)",
+//   "traefik.http.routers.65d2fb6d6732-3771c85af16a-81.service": "65d2fb6d6732-3771c85af16a-81",
+//   "traefik.http.services.65d2fb6d6732-3771c85af16a-81.loadbalancer.server.port": "81"
+func (provider *dockerEnclaveObjectAttributesProviderImpl) getTraefikLabelsForEnclaveObject(serviceUuid string, ports map[string]*port_spec.PortSpec) (map[*docker_label_key.DockerLabelKey]*docker_label_value.DockerLabelValue, error) {
+	labels := map[*docker_label_key.DockerLabelKey]*docker_label_value.DockerLabelValue{}
+
+	for _, portSpec := range ports {
+		maybeApplicationProtocol := ""
+		if portSpec.GetMaybeApplicationProtocol() != nil {
+			maybeApplicationProtocol = *portSpec.GetMaybeApplicationProtocol()
+		}
+		if maybeApplicationProtocol == consts.HttpApplicationProtocol {
+			shortEnclaveUuid := uuid_generator.ShortenedUUIDString(provider.enclaveId.GetString())
+			shortServiceUuid := uuid_generator.ShortenedUUIDString(serviceUuid)
+			servicePortStr := fmt.Sprintf("%s-%s-%d", shortEnclaveUuid, shortServiceUuid, portSpec.GetNumber())
+
+			// Host rule
+			ruleKeySuffix := fmt.Sprintf("http.routers.%s.rule", servicePortStr)
+			ruleLabelKey, err := docker_label_key.CreateNewDockerTraefikLabelKey(ruleKeySuffix)
+			if err != nil {
+				return nil, stacktrace.Propagate(err, "An error occurred getting the traefik rule label key with suffix '%v'", ruleKeySuffix)
+			}
+			ruleValue := fmt.Sprintf("Host(`%d-%s-%s`)", portSpec.GetNumber(), shortServiceUuid, shortEnclaveUuid)
+			ruleLabelValue, err := docker_label_value.CreateNewDockerLabelValue(ruleValue)
+			if err != nil {
+				return nil, stacktrace.Propagate(err, "An error occurred creating the traefik rule label value with value '%v'", ruleValue)
+			}
+			labels[ruleLabelKey] = ruleLabelValue
```

Comment:
@laurentluce I know it's hard to unit test this traefik integration but I wonder if we could break each of these out as functions (host rule, service name  and service port number ) and write some unit tests to verify that they work as intended. It helps lock down the functionality and ensure future engineers have something to test against if they need to change.

Reply:
Alternatively just test the entire function if that's easier?

Reply:
I added a test validating the user service labels including the traefik labels.

Reply:
I am also planning to add integration tests once Traefik is included.